hazelcast/src/main/java/com/hazelcast/client/impl/protocol/ClientMessageReader.java

```diff
@@ -63,13 +63,15 @@ public void reset() {
     }
 
     private boolean readFrame(ByteBuffer src, boolean trusted) {
-        // init internal buffer
-        int remaining = src.remaining();
-        if (remaining < SIZE_OF_FRAME_LENGTH_AND_FLAGS) {
-            // we don't have even the frame length and flags ready
-            return false;
-        }
         if (readOffset == -1) {
+            // Check for the minimum buffer size only if we
+            // haven't read the frame length and flags.
+            int remaining = src.remaining();
+            if (remaining < SIZE_OF_FRAME_LENGTH_AND_FLAGS) {
+                // we don't have even the frame length and flags ready
+                return false;
```

Comment:
If we return here, in which case exception will be thrown at line 77? I'm confused if at the first read, the `remaining` less than 6 bytes, how we expect `frameLength` will be also less than 6 bytes? Is there any case that we throw `IllegalArgumentException`?

Reply:
The first check is to make sure that there are at least 6 bytes we read from the socket so that we can read the frame length and flags.

The second check is there to ensure that the value we read from the first 4 bytes of that 6 bytes (describing the length of the frame) is not an illegal value. Converting the 4 bytes to an integer should yield a value greater than 6 due the protocol.

So, both checks are necessary